src/services/refreshToken.ts

```diff
@@ -0,0 +1,68 @@
+import axios, { AxiosError } from 'axios';
+
+import { refreshToken as authenticationRefreshToken } from 'adapters/Authentication';
+import { authenticatedHeader } from 'adapters/BaseAuth';
+import { clearToken, getToken, setToken } from 'helpers/authentication';
+import { DeserializableResponse, deserialize } from 'helpers/deserializer';
+import { paths } from 'routes';
+import { Token } from 'types/token';
+
+let isAlreadyFetchingAccessToken = false;
+let subscribers: { (isFetchSuccess: boolean): void }[] = [];
+
+function onAccessTokenFetched(isFetchSuccess: boolean) {
+  subscribers = subscribers.filter((callback) => callback(isFetchSuccess));
+}
+
+function addSubscriber(callback: { (isFetchSuccess: boolean): void }) {
+  subscribers.push(callback);
+}
+
+const doRefreshToken = (token: string) => {
+  return authenticationRefreshToken(token)
+    .then((response: DeserializableResponse) => {
+      const tokenType = deserialize<Token>(response.data);
+
+      setToken(tokenType);
+      return true;
```

Comment:
Why do we need these returns?

Reply:
@nvminhtue In order to avoid the infinite calls regarding 401 API, it is important to implement a flag to prevent any potential token refresh failures. That's the purpose of these returns 😄